pkg/skaffold/instrumentation/trace_test.go

```diff
@@ -0,0 +1,159 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package instrumentation
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"os"
+	"testing"
+
+	"github.com/GoogleContainerTools/skaffold/testutil"
+)
+
+func TestInitCloudTrace(t *testing.T) {
+	tests := []struct {
+		shouldError        bool
+		traceProviderIsNil bool
+		name               string
+		traceEnvVar        string
+		stdoutEnvVar       string
+		spans              []string
+		fileSystem         *testutil.FakeFileSystem
+	}{
+		{
+			name: "SKAFFOLD_TRACE=gcp-skaffold and key present, trace provider is not nil",
+			fileSystem: &testutil.FakeFileSystem{
+				Files: map[string][]byte{"/secret/keys.json": []byte(testKey)},
+			},
+			traceEnvVar: "gcp-skaffold",
+		},
+		{
+			name: "SKAFFOLD_TRACE=gcp-skaffold and key not present, returns nil err",
+			fileSystem: &testutil.FakeFileSystem{
+				Files: map[string][]byte{},
+			},
+			traceEnvVar:        "gcp-skaffold",
+			traceProviderIsNil: true,
+		},
+		{
+			name: "SKAFFOLD_TRACE=gcp-skaffold and key present, credentials without project_id returns an error",
+			fileSystem: &testutil.FakeFileSystem{
+				Files: map[string][]byte{
+					"/secret/keys.json": []byte(`{
+						"client_id": "test_id",
+						"client_secret": "test_secret",
+						"refresh_token": "test_token",
+						"type": "authorized_user"
+					}`,
+					)},
+			},
+			traceEnvVar:        "gcp-skaffold",
+			shouldError:        true,
+			traceProviderIsNil: true,
+		},
+		{
+			name:        "SKAFFOLD_TRACE=stdout, verify all spans output to stdout",
+			traceEnvVar: "stdout",
+			spans:       []string{"SpanOne", "SpanTwo"},
+		},
+		{
+			name:         "SKAFFOLD_EXPORT_TO_STDOUT=true, verify all spans output to stdout",
+			stdoutEnvVar: "true",
+			spans:        []string{"SpanOne", "SpanTwo"},
+		},
+	}
+	for _, test := range tests {
+		testutil.Run(t, test.name, func(t *testutil.T) {
+			if len(test.traceEnvVar) > 0 {
+				os.Setenv("SKAFFOLD_TRACE", test.traceEnvVar)
+				defer os.Unsetenv(("SKAFFOLD_TRACE"))
+			}
+			if len(test.stdoutEnvVar) > 0 {
+				os.Setenv("SKAFFOLD_EXPORT_TO_STDOUT", test.stdoutEnvVar)
+				defer os.Unsetenv(("SKAFFOLD_EXPORT_TO_STDOUT"))
+			}
+			var b bytes.Buffer
+			func() {
+				ctx := context.Background()
+				tp, _, err := InitTraceFromEnvVar(WithWriter(&b), WithKeyFS(test.fileSystem))
+				t.CheckErrorAndDeepEqual(test.shouldError, err, test.traceProviderIsNil || test.shouldError, tp == nil)
+				defer func() { _ = TracerShutdown(ctx) }()
+
+				for _, name := range test.spans {
+					_, endTrace := StartTrace(ctx, "trace-test", name)
+					endTrace()
+				}
+			}()
+			if len(test.spans) > 0 {
+				var spans SpanArray
+				err := json.Unmarshal(b.Bytes(), &spans)
+				if err != nil {
+					t.Errorf("unexpected error occurred unmarshalling trace spans %v: %v", b.String(), err)
+				}
+				t.CheckTrue(len(spans) == len(test.spans))
```

Comment:
Don't you want to compare the spans themselves?

Reply:
This part of the test verifies that when I define in the test I want X spans (X StartTrace calls) I indeed see X spans w/ the stdout exporter.  I could do a deep equal on the spans but I would need to mock more otel state and otel itself tests.  I am happy to make this change, I didn't know how useful it would be as it would mirror the tests done in otel itself:
https://github.com/open-telemetry/opentelemetry-go/blob/main/exporters/stdout/trace_test.go

Reply:
Maybe you could just compare what's output on `b`?  I'd like to verify a few situations: sequential spans are properly ordered, child spans are associated to a parent.  We could be messing up with the contexts, or how we call into the OpenTelemtry SDK.

Reply:
Added tests to verify all suggested cases:
sequential spans are properly ordered, concurrent spans work as expected, & child spans are associated to a parent